Can turn immutable into mut with `ref mut`

This lets you get a mutable reference to an immutable value under a specific case.

Code:
```Rust
enum Foo {
	Bar(i32),
}

fn main() {
	let arr = vec!(Foo::Bar(0));
	if let (Some(Foo::Bar(ref mut val)), _) = (&arr.get(0), 0) {
		*val = 9001;
	}
	match arr[0] {
		Foo::Bar(ref s) => println!("{}", s)
	}
}
```

From what I can tell, the tuple and at least two depths of Enums are necessary, along with an `&` on the right side.  There's probably other variations, but that seems to work well enough.

Expected output would be a compile error, instead output is `9001`

This works on both stable `rustc 1.27.1 (5f2b325f6 2018-07-07)` and nightly `rustc 1.29.0-nightly (e5f6498d3 2018-07-10)`

Yes #52232 correctly fixed this issue.

```console
$ rustup-toolchain-install-master 4700e1188f66fdb4086b7593416b678b8fe935f4
downloading <https://s3-us-west-1.amazonaws.com/rust-lang-ci2/rustc-builds/4700e1188f66fdb4086b7593416b678b8fe935f4/rustc-nightly-x86_64-apple-darwin.tar.xz>...
completed                                                                                                                                                                                                                                  
downloading <https://s3-us-west-1.amazonaws.com/rust-lang-ci2/rustc-builds/4700e1188f66fdb4086b7593416b678b8fe935f4/rust-std-nightly-x86_64-apple-darwin.tar.xz>...
completed                                                                                                                                                                                                                                  
toolchain `4700e1188f66fdb4086b7593416b678b8fe935f4` is successfully installed!

$ rustc +4700e1188f66fdb4086b7593416b678b8fe935f4 1.rs
error[E0596]: cannot borrow field of immutable binding as mutable
 --> 1.rs:7:24
  |
7 |     if let (Some(Foo::Bar(ref mut val)), _) = (&arr.get(0), 0) {
  |                           ^^^^^^^^^^^ cannot mutably borrow field of immutable binding

error: aborting due to previous error

For more information about this error, try `rustc --explain E0596`.
```

--- a/src/test/ui/issues/issue-52240.nll.stderr
+++ b/src/test/ui/issues/issue-52240.nll.stderr
@@ -1,0 +1,9 @@
+error[E0596]: cannot borrow data in a `&` reference as mutable
+  --> $DIR/issue-52240.rs:9:27
+   |
+LL |     if let (Some(Foo::Bar(ref mut val)), _) = (&arr.get(0), 0) {
+   |                           ^^^^^^^^^^^ cannot borrow as mutable
+
+error: aborting due to previous error
+
+For more information about this error, try `rustc --explain E0596`.

--- a/src/test/ui/issues/issue-52240.rs
+++ b/src/test/ui/issues/issue-52240.rs
@@ -1,0 +1,16 @@
+// issue-52240: Can turn immutable into mut with `ref mut`
+
+enum Foo {
+    Bar(i32),
+}
+
+fn main() {
+    let arr = vec!(Foo::Bar(0));
+    if let (Some(Foo::Bar(ref mut val)), _) = (&arr.get(0), 0) {
+        //~^ ERROR cannot borrow field of immutable binding as mutable
+        *val = 9001;
+    }
+    match arr[0] {
+        Foo::Bar(ref s) => println!("{}", s)
+    }
+}

--- a/src/test/ui/issues/issue-52240.stderr
+++ b/src/test/ui/issues/issue-52240.stderr
@@ -1,0 +1,9 @@
+error[E0596]: cannot borrow field of immutable binding as mutable
+  --> $DIR/issue-52240.rs:9:27
+   |
+LL |     if let (Some(Foo::Bar(ref mut val)), _) = (&arr.get(0), 0) {
+   |                           ^^^^^^^^^^^ cannot mutably borrow field of immutable binding
+
+error: aborting due to previous error
+
+For more information about this error, try `rustc --explain E0596`.

--- a/src/test/ui/issues/issue-54966.rs
+++ b/src/test/ui/issues/issue-54966.rs
@@ -1,0 +1,6 @@
+// issue-54966: ICE returning an unknown type with impl FnMut
+
+fn generate_duration() -> Oper<impl FnMut()> {}
+//~^ ERROR cannot find type `Oper` in this scope
+
+fn main() {}

--- a/src/test/ui/issues/issue-54966.stderr
+++ b/src/test/ui/issues/issue-54966.stderr
@@ -1,0 +1,9 @@
+error[E0412]: cannot find type `Oper` in this scope
+  --> $DIR/issue-54966.rs:3:27
+   |
+LL | fn generate_duration() -> Oper<impl FnMut()> {}
+   |                           ^^^^ not found in this scope
+
+error: aborting due to previous error
+
+For more information about this error, try `rustc --explain E0412`.

